File: src/univariate/arcsine.jl
# TODO: Implement standard Arcsine(0, 1) and Arcsine(a, b)

immutable Arcsine <: ContinuousUnivariateDistribution
end

@continuous_distr_support Arcsine 0.0 1.0

function cdf(d::Arcsine, x::Real)
    x < zero(x) ? 0.0 : (x > one(x) ? 1.0 : (2.0 / pi) * asin(sqrt(x)))
end

# entropy(d::Arcsine) = log(pi) + digamma(0.5) + eulergamma
# calculated using higher-precision arithmetic 
entropy(d::Arcsine) = -0.24156447527049044469

kurtosis(d::Arcsine) = -1.5

mean(d::Arcsine) = 0.5

median(d::Arcsine) = 0.5

function mgf(d::Arcsine, t::Real)
    s = 0.0
    for k in 1:10
        inner_s = 1.0
        for r in 0:(k - 1)
            inner_s *= (2.0 * r + 1.0) / (2.0 * r + 2.0)
        end
        s += t^k / factorial(k) * inner_s
    end
    1.0 + s
end

function cf(d::Arcsine, t::Real)
    error("CF for Arcsine requires confluent hypergeometric function")
end

mode(d::Arcsine) = 0.0
modes(d::Arcsine) = [0.0, 1.0]

function pdf(d::Arcsine, x::Real)
    zero(x) <= x <= one(x) ? 1.0 / (pi * sqrt(x * (1.0 - x))) : 0.0
end

quantile(d::Arcsine, p::Real) = sin((pi * p) / 2.0)^2

rand(d::Arcsine) = sin(rand() * pi / 2.0)^2

skewness(d::Arcsine) = 0.0

var(d::Arcsine) = 1.0 / 8.0

File: test/ref/continuous/arcsine.R

Arcsine <- R6Class("Arcsine",
    inherit = ContinuousDistribution,
    public = list(
        names = c("a", "b"),
        a = NA,
        b = NA,
        rd = NA,  # R distr object
        initialize = function(a1=NA, a2=NA) {
            if (is.na(a1)) {
                a <- 0; b <- 1
            } else if (is.na(a2)) {
                a <- 0; b <- a1
            } else {
                a <- a1; b <- a2
            }
            self$a <- a
            self$b <- b
            self$rd <- distr::Arcsine() * ((b-a)/2) + ((a+b)/2)
        },
        supp = function() { c(self$a, self$b) },
        properties = function() {
            s <- self$b - self$a
            list(location=self$a,
                 scale=s,
                 mean=(self$a + self$b) * 0.5,
                 var=1/8 * s^2,
                 skewness=0,
                 kurtosis=-1.5,
                 median=(self$a + self$b) * 0.5,
                 entropy=log(pi/4) + log(s))
        },
        pdf = function(x, log=FALSE) { distr::d(self$rd)(x, log=log) },
        cdf = function(x) { distr::p(self$rd)(x) },
        quan = function(v) { distr::q(self$rd)(v) }
    )
)
